src/Microsoft.AspNetCore.Server.HttpSys/RequestProcessing/HeaderCollection.cs

```diff
@@ -4,12 +4,17 @@
 using System;
 using System.Collections;
 using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
 using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
 
 namespace Microsoft.AspNetCore.Server.HttpSys
 {
-    internal class HeaderCollection : IDictionary<string, StringValues>
+    internal class HeaderCollection : IHeaderDictionary
     {
+        private long? _contentLength;
+        private StringValues _contentLengthText;
```

Comment:
We check it against the raw header StringValues and this avoids the extra conversions